### PR TITLE
Hyperchains parent manager

### PR DESCRIPTION
**Pull request by erlmachinedev**
This PR contains the next described functionality + test SUITE which supplied with a simple parent chain demo generator:

### Hyperchains parent's process manager
See the pattern "ProcessManager (https://www.enterpriseintegrationpatterns.com/patterns/messaging/ProcessManager.html).

This component is responsible for orchestration the abstract parent's chain layer through:
 a) dedicated workers (trackers);
 b) supplied interface providers (connectors);

### Notes
 - The first supplied connector acts as the "master" and is eligible to dictate the election period through the event's announcement;
 - A tracker performs synchronization from the highest known block until the genesis pointer through the "previous_hash" property;
 - To be able to run the Hyperchains the system must satisfy the connector's acceptance criteria;
     a) default mode: get_top_block/0, get_block_by_hash/1l
     b) delegate mode: default mode + send_tx/1;
 - Each interested developer can supply their own connector's implementation over the particular parent chain;

### CODE REVIEW

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+        args :: map(),
+        %% Textual description;
+        note :: binary()
+    }).
+
+-type track() :: #track{}.
+%% API.
+
+-spec start_link() ->
+    {ok, pid()} | ingnore | {error, term()}.
```
Comment (radrow) [inline apps/aehyperchains/src/aehc_parent_mng.erl:63]:
```suggestion
    {ok, pid()} | ignore | {error, term()}.
```

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    State#state{tracks = Update}.
+
+%%%===================================================================
+%%%  Connector's management
+%%%===================================================================
+
+-spec accept_connector(track()) -> boolean().
+accept_connector(Track) ->
+    try
+        [Criteria(Track) || Criteria <- fun accept_top_block/1, fun accept_block_by_hash/1, fun accept_send_tx/1],
```
Comment (radrow) [inline apps/aehyperchains/src/aehc_parent_mng.erl:149]:
```suggestion
        [Criteria(Track) || Criteria <- [fun accept_top_block/1, fun accept_block_by_hash/1, fun accept_send_tx/1]],
```

**Review by radrow — CHANGES_REQUESTED** [review]

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    {ok, _} = aehc_connector:get_block_by_hash(Con, Genesis).
+
+accept_send_tx(Track) ->
+    %% Ability to execute commitment call;
+    delegate_config() == undefined orelse
+        begin
+            Con = connector(Track),
+            Delegate = aec_keys:pubkey(),
+            KeyblockHash = aec_chain:top_key_block_hash(),
+            Payload = aehc_commitment_header:hash(aehc_commitment_header:new(Delegate, KeyblockHash)),
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:180]: You don't submit the commitment hash

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+            ok = aehc_connector:send_tx(Con, Payload)
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:181]: I don't like this sanity check

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+                GenesisHash;
+            _ when is_binary(Res) ->
+                Res
+        end,
+    Block = aehc_parent_db:get_parent_top_block(TopBlockHash),
+    init_track(Track, Block).
+
+%% Sync of track by fetching the current top;
+-spec sync_track(track()) -> track().
+sync_track(Track) ->
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:206]: This is blocking...

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    %% Read configuration;;
+    Tracks = [conf_track(C) || C <- tracks_config()],
+    %% Db initialization
+    InitTracks = [init_db_track(Track) || Track <- Tracks],
+    %% Execute acceptance procedure;
+    [accept_connector(T) || T <- InitTracks],
+    %% Run conenctor's instances;
+    [start_connector(T) || T <- InitTracks],
+    %% Forks synchronization by fetched blocks;
+    SynchedTracks = [sync_track(T) || T <- InitTracks],
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:91]: This is blocking... This needs to be done in separate processes

**Review by gorbak25 — CHANGES_REQUESTED** [review]
To summarize:
1. More tests, perhaps develop more in line with TDD.
2. Use monitors... instead of trapping exits...
3. Tracks should represent "A sync job which is currently under way"
4. Each track should be processed in it's own process - coordinated by the connector manager which provides an API for the currently active jobs
5. The master connector should not block for a week in the init function - just generate some initial jobs,  monitor the job pids and enter the genserver loop...
6. Sync tasks should be parallelized - The manager should keep track of the amount of jobs for a given connector, If a given sync job(track) can be split in half the do it if we have a small ammout of workers currently active.

7. Besides that good job! Unfortunately this PR still requires a lot of work.

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 7]
```diff
@@ -0,0 +1,333 @@
+%%% -*- erlang-indent-level: 4 -*-
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2020, Aeternity Anstalt
+%%% @doc
+%%% Hyperchains parent's layer manager
+%%% This component is responsible to manage the abstract parent's chain layer through so called "tracks".
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:6]: Thanks

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+%%% Each track consists of a synchronized address at the particular parent chain and is supplied with a responsible interface provider.
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:7]: Thanks

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+%%% Please note that:
+%%% - The first track supplies the "master" connector and eligible to dictate the election period through event's announcement;
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:9]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+%%% - The synchronization occurs from the current highest block until track's genesis_hash value has achived;
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:10]: Thanks

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+%%% - Here is and the next time we mention term "genesis hash" (please don't confuse with the common "genesis block" terminology);
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:11]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+%%% - The component is responsible to manage fork switching by traversing hash-pointers of the parent blocks. Each time when fork occurs:
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:12]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+%%%     a) The database is updated by the new parent chain representation;
+%%%     b) The child chain is reverted and the new election process is started from the last synced track height;
+%%% - To be able to run the Hyperchains the system must satisfy the connector's acceptance criteria;
+%%%     a) For default mode: get_top_block/0, get_block_by_hash/1l
+%%%     b) For delegate mode: default mode + send_tx/1;
+%%% - Each interested developer can supply his own connector's implementation for the particular parent chain;
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:18]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl (hunk 1 above)
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:63]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl (hunk 2 above)
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:149]: Oh, thanks

File: apps/aehyperchains/src/aehc_connector.erl [hunk 14]
```diff
@@ -2,72 +2,64 @@
...
--spec block(Hash::binary(), PrevHash::binary(), Txs::[tx()]) -> block().
-block(Hash, PrevHash, Txs) when
-      is_binary(Hash), is_binary(PrevHash), is_list(Txs) ->
-    #block{ hash = Hash, prev_hash = PrevHash, txs = Txs }.
+%% This API should be used by connector's developer as a wrapper around internal abstract commitments data type;
+-spec commitment(Delegate::binary(), KeyblockHash::binary(), PoGFHash::binary()) ->
+    commitment().
+commitment(Delegate, KeyblockHash, PoGFHash) when
+    is_binary(Delegate), is_binary(KeyblockHash), is_binary(PoGFHash) ->
+    aehc_commitment:new(aehc_commitment_header:new(Delegate, KeyblockHash, PoGFHash)).
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:29]: I faced some issue, have to disscuss. Our API allows to create by:

`-spec new(aec_headers:key_header(), aec_headers:key_header()) -> pogf().`

But we have to be able to reproduce PoGF object from payload record as is, I mean something like:

`-spec new(binary()), binary())) -> pogf().`

Moreover, aec_headers:key_header() entries duplicate data from the current existing ones (already persisted on blockchain).

I propouse to reduce type pogf() into hash storing document

File: apps/aehyperchains/src/aehc_parent_block.erl [hunk 15]
```diff
@@ -78,6 +80,16 @@ commitments_in_block(#hc_parent_block{commitments = Commitments}) ->
...
 
+-spec height_block(parent_block()) -> binary().
+height_block(#hc_parent_block{header = #hc_parent_block_header{height = Height}}) ->
+    Height.
+
 -spec block_header(parent_block()) -> parent_block_header().
 block_header(#hc_parent_block{header = Header}) ->
     Header.
+
+-spec is_hc_parent_block(parent_block()) -> boolean().
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_block.erl:91]: Sorry, can't see what is wrong here. This function acts as a filter within manager subsribtion duties, to check arrived message for correctness from a hight level type perspective.

File: apps/aehyperchains/src/aehc_app.erl [hunk 16]
```diff
@@ -26,7 +22,7 @@ stop(_State) ->
     ok.
 
 check_env() ->
-    check_env([{[<<"chain">>, <<"hyperchains">>, <<"enabled">>], {set_env, enabled}}]),
+    check_env([{[<<"hyperchains">>, <<"enabled">>], {set_env, enabled}}]),
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_app.erl:25]: Thanks, also think that hyperchains section should be placed at higher level

File: apps/aehyperchains/src/aehc_parent_mng.erl (hunk 4 above)
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:181]: As you will wish, sir

File: apps/aehyperchains/src/aehc_commitment.erl [hunk 17]
```diff
@@ -49,3 +59,20 @@ header(#hc_commitment{header = Header}) ->
...
+payload_delegate(<<Delegate:?COMMITER_PUB_BYTES, _:?COMMITMENT_HASH_BYTES, _:?POGF_HASH_BYTES>>) ->
+    Delegate.
+
+
+-spec payload_commitment(commitment_hash()) -> block_header_hash().
+payload_commitment(<<_:?COMMITER_PUB_BYTES, Commitment:?COMMITMENT_HASH_BYTES, _:?POGF_HASH_BYTES>>) ->
+    Commitment.
+
+-spec payload_pogf(commitment_hash()) -> pogf_hash().
+payload_pogf(<<_:?COMMITER_PUB_BYTES, _:?COMMITMENT_HASH_BYTES, PoGFHash:?POGF_HASH_BYTES>>) ->
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_commitment.erl:77]: I can explain the need. The connector module should be able to operate payload data (at least to provide signature check we have to extract **"delegate"** piece of data). The intention is to provide payload accessor which can take binary payload as an argument and to extract particular data from it.

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+%%%===================================================================
+
+%% TODO: Connector check API;
+-record(state, { master :: aehc_connector:connector(), tracks :: list() }).
+
+init([]) ->
+    %% TODO: To extract the current block from Db;
+    process_flag(trap_exit, true),
+    %% Read configuration;;
+    Tracks = [conf_track(C) || C <- tracks_config()],
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:83]: Oh, you know I appreciate your intention to improve (that's great quality). 
But, please note, that after we wrapped Db API into transaction calls - that all doesn't matter. I hope you know how transaction mechanism works, we work around database and eventially our DB (and we are too) is sync.

File: apps/aehyperchains/src/aehc_parent_mng.erl (hunk 6 above)
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:90]: Agree, manager should operate independent processes, sure

File: apps/aehyperchains/src/aehc_parent_mng.erl (hunk 5 above)
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:208]: Yep, you are right, but:

1) Are we eligible to proceed until one of the connectors isn't ready? Sounds simple, but actually I am not sure about;
2) I mention you transaction consequences below. We work around database and this table each time in a "freeze" state until call has executed. That means that a lot of parallelizing features don't impact here.

So, in general I agree that we have to decouple job into processes, and that's clear for me. But please also note the sync nature of some init procedures which we can't skip.

File: apps/aehyperchains/src/aehc_parent_mng.erl (hunk 18 above)
[the remark this answers is not in the document]
Comment (gorbak25) [reply apps/aehyperchains/src/aehc_parent_mng.erl]: Your argument is valid ONLY when the time to retrieve data from the parent chain is negligible which might be the case for our demo "chain simulator" connector.
In the real world the time to retrieve blocks and commitments from the parent chain is significant and can't be ignored. Spiting up tracks will yield an increase in performance for:
- Connectors using HTTP to retrieve data from an API endpoint
- Connectors which are essentially light clients and validate data provided by other peers

Anyway - if we got an async job infrastructure setup then why not parallelize the jobs?

File: apps/aehyperchains/src/aehc_parent_mng.erl (hunk 5 above)
Comment (gorbak25) [reply apps/aehyperchains/src/aehc_parent_mng.erl]: Other parts of the node might want to inspect the status of the parent sync - especially a user wondering why the node isn't fully booting. Without entering the genserver loop quickly we essentially refuse to respond to requests about the status of the sync jobs. Please take a look at aec_sync which has a similar problem.

After sync is done the mining conductor should get notified that "Hey we are ready to participate in elections".

File: apps/aehyperchains/src/aehc_connector.erl [hunk 19]
```diff
@@ -2,96 +2,138 @@
...
+%% TODO: Opened for discussion with gorbak25 and radrow;
+%%-spec commitment(Delegate::binary(), KeyblockHash::binary(), PoGFHash::binary()) ->
+%%    commitment().
+%%commitment(Delegate, KeyblockHash, PoGFHash) when
+%%    is_binary(Delegate), is_binary(KeyblockHash), is_binary(PoGFHash) ->
+%%    aehc_commitment:new(aehc_commitment_header:new(Delegate, KeyblockHash, PoGFHash)).
+
+-spec parent_block(Height::non_neg_integer(), Hash::binary(), PrevHash::binary(), Commitments::[commitment()]) ->
+    parent_block().
+parent_block(Height, Hash, PrevHash, Commitments) when
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_connector.erl:44]: Ok, this helper is still broken and generates an invalid parent block ADT.
Please preserve the following invariant:
```
is_list(Block.Header.commitment_hashes),
is_list(Block.commitments),
length(Block.Header.commitment_hashes) == length(Block.commitments),
[Hash = aehc_commitment:hash(Commitment) || {Hash, Commitment} <- lists:zip(Block.Header.commitment_hashes, Block.commitments)]
```
This is essentially the same invariant we preserve when handling microblocks. 
Take a look at:
https://github.com/aeternity/aeternity/blob/8b33ad36d6f81d90f57178b735819c01bfe652c8/apps/aehyperchains/src/aehc_parent_db.erl#L96

File: apps/aehyperchains/src/aehc_parent_db.erl [hunk 20]
```diff
@@ -143,3 +150,24 @@ write_parent_block(ParentBlock) ->
...
+
+%% The parent chain view entry determines traversing path within the current persisted log (genesis -> top);
+-spec write_parent_chain_view(binary(), binary()) -> ok.
+write_parent_chain_view(GenesisHash, TopHash) when is_binary(GenesisHash),
+                                                    is_binary(TopHash) ->
+    ?t(mnesia:write(#hc_db_parent_chain_view{key = GenesisHash, value = TopHash}),
+        [{hc_db_parent_chain_view, GenesisHash}]).
+
+-spec get_parent_chain_view(binary()) -> binary() | undefined.
+get_parent_chain_view(GenesisHash) ->
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_db.erl:162]:
```suggestion
find_parent_chain_view(GenesisHash) ->
```

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+%%% <h3>Hyperchains parent's process manager</h3>
+%%% See the pattern "ProcessManager" (https://www.enterpriseintegrationpatterns.com/patterns/messaging/ProcessManager.html).
+%%% This component is responsible for orchestration the abstract parent's chain layer through:
+%%%  a) dedicated workers (trackers);
+%%%  b) supplied interface providers (conenctors);
+%%% - The first configured view acts as the "master view" and is eligible to dictate the election period through event's announcement;
+%%% - A tracker performs synchronization from the highest known block until genesis pointer through "previous_hash" property;
+%%% - To be able to run the Hyperchains the system must satisfy the connector's acceptance criteria;
+%%%     a) default mode: get_top_block/0, get_block_by_hash/1l
+%%%     b) delegate mode: default mode + send_tx/1;
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:14]: send_tx shouldn't send out a real tx when initializing the system

File: apps/aehyperchains/src/aehc_parent_mng.erl [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+handle_call({terminate_view, View}, _From, State) ->
+    Res = aehc_sup:terminate_view(View),
+    lager:info("~p terminate parent view: ~p", [View]),
+    {reply, Res, State}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info({publish_block, View, Block}, #state{} = State) ->
+    %% To be able to persists the block we have to be sure that it connects to the previous existed one in the DB;
```
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:85]:
```suggestion
    %% To be able to persists the block we have to be sure that it connects to the previous existing one in the DB;
```

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl [hunk 23]
```diff
@@ -46,22 +46,22 @@ get_block_by_hash(Hash) ->
...
-    {ok, Pub} = aec_keys:pubkey(),
     {ok, PrivKey} = aec_keys:sign_privkey(),
-    SenderId = aeser_id:create(account, Pub),
     %% The main intention of this call is to emulate post action with signed payload from delegate;
     %% Fee, nonce, ttl and amount fields have decorated nature;
-    {ok, Tx} = aec_spend_tx:new(#{ sender_id => SenderId, recipient_id => SenderId, amount => 1,
+    Header = aehc_commitment_header:new(<<"D1">>, <<"BLOCK 1">>),
+    Payload = aehc_commitment_header:hash(Header),
+    {ok, Tx} = aec_spend_tx:new(#{ sender_id => Delegate, recipient_id => Delegate, amount => 1,
                                         fee => 5, nonce => 1, payload => Payload, ttl => 0 }),
```
Comment (gorbak25) [inline apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:65]: You are sending the commitment ID instead of the contents.

**Review by gorbak25 — CHANGES_REQUESTED** [review]
I like the direction this PR is moving but some small work still needs to be done before it can be merged:
- Only empty blocks are properly saved in the DB - the code fails for blocks with commitments
- aehc_parent_mng:start_view should be called by aehc_parent_mng and not the user
- Improper usage of commitment hashes - a commitment hash is a unique identifier not the contents we send out to the parent chain

File: apps/aehyperchains/src/aehc_connector.erl (hunk 19 above)
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:44]: How much beer have you drunk yesterday that posted Elixir syntax here? Joke :) Missed one argument in constructor, thanks for the notice.

File: apps/aehyperchains/src/aehc_connector.erl [hunk 24]
```diff
@@ -2,96 +2,138 @@
...
+    is_integer(Height), is_binary(Hash), is_binary(PrevHash), is_list(Commitments) ->
+    Header = aehc_parent_block:new_header(Hash, PrevHash, Height),
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:46]: **1, 3** - got you, will fix today
About **2**, just short notice to be sure that we at "the same ground here". The main design consideration around parent_mng is the ability to manage the interaction between multiple parent chains, like:

1) Who dictates the election period?
2) What we can do in a conflict situation that arose between multiple parent views?
3) Some cases can come later and this component can be improved by additional functionality;

So it's not a supervisor, for the monitor purposes we have - supervisor :). API is called by our app and is exported with keeping in mind the dynamic nature of HC - when something bad happens outside the operator should be able to switch or reject some chain.

File: apps/aehyperchains/src/aehc_connector.erl [hunk 25]
```diff
@@ -2,96 +2,138 @@
...
+%%     Ability to request genesis block;
+    Module = module(Conf),
+    Genesis = maps:get(<<"genesis_hash">>, Conf),
+    {ok, Block} = get_block_by_hash(Module, Genesis),
+    Hash = aehc_parent_block:hash_block(Block),
+    Height = aehc_parent_block:height_block(Block),
+    Info = "Accept get_block_by_hash procedure has passed (connector: ~p, hash: ~p, height: ~p)",
+    lager:info(Info, [Module, Hash, Height]).
+
+accept_send_tx(Conf) ->
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:111]: Maybe you are right, just over measure from me. Idea with dry_send_tx seems ok

[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:111]: We perform now two checks: get_top_block(), get_block_by_hash(Genesis), and both of them are independent and have to be satisfied;

Implementation of dry_send_tx can be delegated to a particular connector:

dry_send_tx -> just to return **ok** (if I am not going to spend fee);
dry_send_tx -> more specific check, like contract existence, available balance etc.

File: apps/aehyperchains/src/aehc_connector.erl [hunk 26]
```diff
@@ -2,96 +2,138 @@
...
 %%%===================================================================
 
-%% TODO: BlockHash, etc..
+-spec publish_block(term(), parent_block()) -> ok.
+publish_block(View, Block) ->
+    aehc_parent_mng:publish_block(View, Block).
+
+module(Conf) ->
+    ConConf = maps:get(<<"connector">>, Conf),
+    binary_to_existing_atom(maps:get(<<"module">>, ConConf), utf8).
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:135]: Today you have that knowledge that "It's a one-time conversion", tomorrow me, or new Adam, or somebody else will use that in another way - and we got node leak.

These rules have written by blood - I am not going to joke here). I will mark this in code with your suggestion.

File: apps/aehyperchains/src/aehc_parent_db.erl [hunk 27]
```diff
@@ -143,3 +150,24 @@ write_parent_block(ParentBlock) ->
            [mnesia:write(DBCommitment) || DBCommitment <- DBCommitments],
            [mnesia:write(DBPoGF) || DBPoGF <- DBPoGFs]
        end).
+
+%% The parent chain view entry determines traversing path within the current persisted log (genesis -> top);
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_db.erl:154]: ok

File: apps/aehyperchains/src/aehc_parent_db.erl (hunk 20 above)
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_db.erl:162]: write_, get_ are convenient prefixes here.

File: apps/aehyperchains/src/aehc_parent_db.erl [hunk 28]
```diff
@@ -143,3 +150,24 @@ write_parent_block(ParentBlock) ->
...
+           [#hc_db_parent_chain_view{value = Value}] ->
+               Value;
+           _ ->
+               undefined
+       end).
+
+%% Get parent tob block within particular parent view;
+-spec get_parent_top_block(binary()) -> aehc_parent_block:parent_block() | undefined.
+get_parent_top_block(GenesisHash) ->
+    ?t(get_parent_block(get_parent_chain_view(GenesisHash))).
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_db.erl:173]: This API reflects aec_chain:

top_block/0 -> get_parent_top_block/0;
get_block/1 -> get_parent_block/1

**Review by gorbak25 — CHANGES_REQUESTED** [review]
Good job! We are close to merging this PR.
2 small things remain before merging this to HC-master:
- The commits will be squashed anyway but please rebase your branch and ensure that the commits are more or less clear (69 commits with multiple merges of hyperchains-master + duplicated commits due to merging instead of rebasing is extremely unclear). It doesn't have to be perfect but at least remove the standing out things - merges, duplicates etc...
- Add a test case for saving a block with commitments via a connector and reading it back from DB - currently we only test blocks without commitments

Optional:
Enable GPG auto signing of your commits.

**Review by gorbak25 — APPROVED** [review]

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl [hunk 29]
```diff
@@ -15,53 +15,59 @@
 -export([handle_info/2]).
 -export([terminate/2]).
 
--export([send_tx/1, get_block_by_hash/1, get_top_block/0]).
+-export([send_tx/3, get_block_by_hash/1, get_top_block/0, dry_send_tx/3]).
 
 %% API.
 
--spec start_link() ->
+-spec start_link(Args::term()) ->
```
Comment (radrow) [inline apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:22]: why so unspecified?

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl (hunk 23 above)
Comment (radrow) [reply apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:71]: +1

File: apps/aehyperchains/src/aehc_parent_block.erl [hunk 30]
```diff
@@ -78,6 +80,16 @@ commitments_in_block(#hc_parent_block{commitments = Commitments}) ->
...
 
+-spec height_block(parent_block()) -> non_neg_integer().
+height_block(#hc_parent_block{header = #hc_parent_block_header{height = Height}}) ->
+    Height.
+
 -spec block_header(parent_block()) -> parent_block_header().
 block_header(#hc_parent_block{header = Header}) ->
     Header.
+
+-spec is_hc_parent_block(parent_block() | term()) -> boolean().
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_block.erl:91]: +

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl (hunk 29 above)
Comment (erlmachinedev) [reply apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:22]: +

File: apps/aehyperchains/src/aehc_parent_block.erl [hunk 31]
```diff
@@ -17,7 +17,9 @@
         , prev_hash_block/1
         , commitments_in_block/1
         , hash_block/1
+        , height_block/1
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_block.erl:20]: prev_hash_block, hash_block, height_block - for the sake of compatibility

File: apps/aehyperchains/src/aehc_connector.erl [hunk 32]
```diff
@@ -2,96 +2,150 @@
...
 
--spec subscribe_block() -> true.
-subscribe_block() ->
-    aec_events:subscribe({parent_chain, block}).
-
--spec publish_block(block()) -> ok.
-publish_block(Block) ->
-    aec_events:publish({parent_chain, block}, {block_created, Block}).
+accept_top_block(Conf) ->
+    %% Ability to request the current top block;
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:103]: It's short notice, works for me (but I got you if you see some mess here I remove)

File: apps/aehyperchains/src/aehc_connector.erl [hunk 33]
```diff
@@ -2,96 +2,150 @@
...
--spec block(Hash::binary(), PrevHash::binary(), Txs::[tx()]) -> block().
-block(Hash, PrevHash, Txs) when
-      is_binary(Hash), is_binary(PrevHash), is_list(Txs) ->
-    #block{ hash = Hash, prev_hash = PrevHash, txs = Txs }.
+%% This API should be used by connector's developer as a wrapper around internal abstract commitments data type;
+-spec commitment(Delegate::binary(), KeyblockHash::binary()) ->
+    commitment().
+commitment(Delegate, KeyblockHash) when
+    is_binary(Delegate), is_binary(KeyblockHash) ->
+    aehc_commitment:new(aehc_commitment_header:new(Delegate, KeyblockHash)).
```
[the remark this answers is not in the document]
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:35]: Yep, you are right - fixed

**Pull request merged**